enterprise_access/apps/api/v1/views/utils.py

```diff
@@ -1,19 +1,59 @@
 """
-from rest_framework.pagination import PageNumberPagination
+from django.conf import settings
+from edx_rest_framework_extensions.paginators import DefaultPagination
 
 
-class PaginationWithPageCount(PageNumberPagination):
+class PaginationWithPageCount(DefaultPagination):
     """
-    A PageNumber paginator that adds the total number of pages to the paginated response.
+    A PageNumber paginator that adds the total number of pages
+    and the current page to the paginated response.
     """
 
     page_size_query_param = 'page_size'
+    # The configured `PAGE_SIZE` in Django settings must be used as the default page size as opposed
+    # to relying on the default `page_size` specified via `DefaultPagination` in order to maintain
+    # backwards compatibility with existing API clients.
+    page_size = settings.REST_FRAMEWORK.get('PAGE_SIZE')
     max_page_size = 500
 
-    def get_paginated_response(self, data):
-        """ Adds a ``num_pages`` field into the paginated response. """
-        response = super().get_paginated_response(data)
-        response.data['num_pages'] = self.page.paginator.num_pages
-        return response
+
+class OptionalPaginationWithPageCount(PaginationWithPageCount):
+    """
+    Optionally allows callers to disable pagination by setting the (configurable) `no_page` query param.
+    """
+    no_page_query_param = 'no_page'
+
+    def formatted_no_page_query_param(self, request):
+        """
+        Parses the query parameters for `self.no_page_query_param` and transforms
+        its value to a lowercase string, if present.
+        """
+        no_page_query_param = request.query_params.get(self.no_page_query_param)
+        formatted_no_page_query_param = no_page_query_param.lower() if no_page_query_param else no_page_query_param
+        return formatted_no_page_query_param
+
+    def has_non_paginated_response(self, request):
+        """
+        Determines whether the given `self.no_page_query_param` query param has a value that indicates
+        that pagination should be disabled for the request.
+        """
+        no_page_query_param = self.formatted_no_page_query_param(request)
+        return no_page_query_param in ('true', '1')
+
+    def paginate_queryset(self, queryset, request, view=None):
+        """
+        Paginates a queryset, if required. Allows disabling pagination by setting
+        the `no_page` query param, supporting use cases where API callers may not be
+        able to rely on pagination (e.g., if more than 1 page of results could be expected
+        for a use case where the client can't traverse all pages and passing a "sufficiently high"
+        `page_size` query parameter is too naive of a solution).
+        """
+        # Determine whether to paginate the queryset based on the value of the
+        # `no_page` query param. If True, do not paginate the response.
+        if self.has_non_paginated_response(request):
+            return None
```

Comment:
[inform] By returning `None` here, DRF never calls `self.get_paginated_response()`, so we wouldn't be able to conditionally modify `self.get_paginated_response()` to behave different with the presence of a `?no_page=1` query parameter.

It feels preferable to rely on this logic living in the paginator than the viewset itself even though it means we have to  return a top-level list, e.g.:

```json
[
    {
        "uuid": "db2372f3-c25c-4c38-8de0-60d4d105e0e3",
        "assignment_configuration": "942410d4-ef20-4e4a-9a5f-fa2da07cbb39",
        "learner_email": "[edx@example.com](mailto:edx@example.com)",
        "lms_user_id": null,
        "content_key": "edX+DemoX",
        "content_quantity": -19900,
        "state": "allocated",
        "transaction_uuid": null,
        "last_notification_at": null
    },
    {
        "uuid": "f9c7c4c8-11a7-45f7-ae3e-d81029ec5541",
        "assignment_configuration": "942410d4-ef20-4e4a-9a5f-fa2da07cbb39",
        "learner_email": "[edx@example.com](mailto:edx@example.com)",
        "lms_user_id": null,
        "content_key": "HarvardX+CS50x",
        "content_quantity": -500000,
        "state": "allocated",
        "transaction_uuid": null,
        "last_notification_at": null
    }
]
```

FWIW, the top-level list for non-paginated responses does seem to be the current convention for several (most?) non-paginated list APIs in Enterprise (e.g., `enterprise_course_enrollments`) and the broader Open edX platform.